File: backend/handlers/member_summary_handler.py
from datetime import datetime
from sqlalchemy import func
from sqlalchemy.orm import Session

from handlers.base import WidgetDataHandler
from models import Issue, Member, Department, Project
from schemas import WidgetQueryRequest, WidgetDataResponse, ColumnDef, ResultMeta

MONTH_KEYS = ["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"]


class MemberMonthlySummaryHandler(WidgetDataHandler):
    @property
    def source_key(self) -> str:
        return "member_monthly_summary"

    def query(self, request: WidgetQueryRequest, db: Session) -> WidgetDataResponse:
        f = request.filters
        year          = (f.year if f else None) or datetime.now().year
        department_id = f.departmentId if f else None

        project = db.query(Project).filter(Project.key == request.projectKey).first()
        if not project:
            raise ValueError(f"Project '{request.projectKey}' not found")

        q = (
            db.query(
                Member.id.label("mid"),
                Member.name.label("mname"),
                Department.name.label("dname"),
                func.strftime("%m", Issue.resolved_at).label("month"),
                func.count().label("cnt"),
            )
            .join(Issue.assignee)
            .join(Member.department)
            .filter(
                Issue.project_id == project.id,
                Issue.resolved_at.isnot(None),
                func.strftime("%Y", Issue.resolved_at) == str(year),
            )
        )
        if department_id:
            q = q.filter(Department.id == department_id)

        pivot: dict[int, dict] = {}
        meta_map: dict[int, tuple] = {}

        for r in q.group_by(Member.id, func.strftime("%m", Issue.resolved_at)).order_by(Member.name).all():
            if r.mid not in meta_map:
                meta_map[r.mid] = (r.mname, r.dname)
            pivot.setdefault(r.mid, {})[MONTH_KEYS[int(r.month) - 1]] = r.cnt

        # Ensure all dept members appear even with 0 issues
        if not pivot:
            mq = db.query(Member).join(Member.department).filter(Department.project_id == project.id)
            if department_id:
                mq = mq.filter(Department.id == department_id)
            for m in mq.all():
                meta_map.setdefault(m.id, (m.name, m.department.name))
                pivot.setdefault(m.id, {})

        rows = []
        for mid, monthly in sorted(pivot.items(), key=lambda x: meta_map[x[0]][0]):
            name, dept = meta_map[mid]
            row = {"member": name, "department": dept}
            total = 0
            for mk in MONTH_KEYS:
                v = monthly.get(mk, 0)
                row[mk] = v
                total += v
            row["total"] = total
            rows.append(row)

        columns = (
            [ColumnDef(key="member", label="Member", type="string"),
             ColumnDef(key="department", label="Department", type="string")]
            + [ColumnDef(key=mk, label=mk, type="number") for mk in MONTH_KEYS]
            + [ColumnDef(key="total", label="Total", type="number")]
        )

        return WidgetDataResponse(
            dataSource=self.source_key,
            type="table",
            columns=columns,
            rows=rows,
            meta=ResultMeta(computedAt=datetime.utcnow().isoformat(), fromCache=False, totalRows=len(rows)),
        )

File: backend-java/src/main/java/com/dashboard/handler/MemberSummaryHandler.java
package com.dashboard.handler;

import jakarta.persistence.EntityManager;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.*;

@Component
public class MemberSummaryHandler {

    private static final String[] MONTH_KEYS = {
        "Jan","Feb","Mar","Apr","May","Jun",
        "Jul","Aug","Sep","Oct","Nov","Dec"
    };

    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> query(String projectKey, Integer year, Long departmentId,
                                           EntityManager em) {
        if (year == null) year = LocalDateTime.now().getYear();
        long deptParam = departmentId != null ? departmentId : -1L;

        String sql = """
            SELECT m.id, m.name,
                   CAST(EXTRACT(MONTH FROM i.resolved_at) AS INTEGER) AS month,
                   COUNT(*) AS cnt
            FROM issues i
            JOIN members m ON i.assignee_id = m.id
            JOIN departments d ON m.department_id = d.id
            WHERE i.project_id = (SELECT id FROM projects WHERE key = :key)
              AND i.resolved_at IS NOT NULL
              AND EXTRACT(YEAR FROM i.resolved_at) = :year
              AND (:deptId = -1 OR d.id = :deptId)
            GROUP BY m.id, m.name, CAST(EXTRACT(MONTH FROM i.resolved_at) AS INTEGER)
            ORDER BY m.name
            """;

        var q = em.createNativeQuery(sql)
                  .setParameter("key", projectKey)
                  .setParameter("year", year)
                  .setParameter("deptId", deptParam);

        Map<Long, String>               metaMap = new LinkedHashMap<>();
        Map<Long, Map<String, Integer>> pivot   = new LinkedHashMap<>();

        for (Object[] row : (List<Object[]>) q.getResultList()) {
            Long   mid   = ((Number) row[0]).longValue();
            String mname = (String) row[1];
            int    month = ((Number) row[2]).intValue();
            int    cnt   = ((Number) row[3]).intValue();

            metaMap.putIfAbsent(mid, mname);
            pivot.computeIfAbsent(mid, k -> new HashMap<>())
                 .put(MONTH_KEYS[month - 1], cnt);
        }

        // If no resolved issues, still show all members in the project/dept
        if (metaMap.isEmpty()) {
            String memberSql = departmentId == null
                ? "SELECT m.id, m.name FROM members m JOIN departments d ON m.department_id = d.id JOIN projects p ON d.project_id = p.id WHERE p.key = :key ORDER BY m.name"
                : "SELECT m.id, m.name FROM members m JOIN departments d ON m.department_id = d.id WHERE d.id = :deptId ORDER BY m.name";

            var mq = em.createNativeQuery(memberSql);
            if (departmentId == null) mq.setParameter("key", projectKey);
            else mq.setParameter("deptId", departmentId);

            for (Object[] row : (List<Object[]>) mq.getResultList()) {
                Long mid = ((Number) row[0]).longValue();
                metaMap.putIfAbsent(mid, (String) row[1]);
                pivot.putIfAbsent(mid, new HashMap<>());
            }
        }

        // Build rows
        List<Map<String, Object>> rows = new ArrayList<>();
        for (Long mid : metaMap.keySet()) {
            Map<String, Integer> monthly = pivot.getOrDefault(mid, Map.of());

            Map<String, Object> row = new LinkedHashMap<>();
            row.put("member", metaMap.get(mid));
            int total = 0;
            for (String mk : MONTH_KEYS) {
                int v = monthly.getOrDefault(mk, 0);
                row.put(mk, v);
                total += v;
            }
            row.put("total", total);
            rows.add(row);
        }

        return rows;
    }
}
